### DISCUSSION TITLE
Working example of store type inference

### DISCUSSION THREAD

**Discussion by shfrmn** [pmndrs/zustand#1714] [Ideas]
Hi, everyone!

I understand that the issue of store type inference has been brought up numerous times and there are already some existing solutions. I have read the documentation regarding `combine` middleware, as well as some other discussion threads, such as [this](https://github.com/pmndrs/zustand/discussions/991) and [this](https://github.com/pmndrs/zustand/discussions/1520). From what I read I also gather there's understandable fatigue with regards to this issue, but please hear me out.

Providing `zustand` would export a `SetState` type, with a minor tweak to accept a function type parameter: 

```typescript
type SetState<T> = (
  partial: T extends (...args: any[]) => any
    ? // T is a function
        | Partial<ReturnType<T>>
        | ((state: ReturnType<T>) => Partial<ReturnType<T>>)
      // T is not a function
    : Partial<T> | ((state: T) => Partial<T>),
  replace?: boolean
) => void
```

The example below would just work out of the box with 100% correct type inference both inside and outside the function.
```typescript
import {create, SetState} from "zustand"

function PostStore(set: SetState<typeof PostStore>) {
  return {
    posts: [] as string[],
    addPost: (post: string) => set(({posts}) => ({posts: [...posts, post]})),
  }
}

export const usePosts = create(PostStore)
```

The only difference from the currently recommended basic usage pattern here is extracting the arrow function from inside `create` into a named function, and adding `SetState` type signature.

**Pros**
- No breaking changes and minimal changes to documentation
- Works with uncurried `create`
- Doesn't interfere with any other usage patterns (e.g. `combine`)
- Not a hacky or complex solution
- Simple folk like me get to do our thing without extra middleware and going through the caveats of documentation and warning regarding ["being a little careful"](https://github.com/pmndrs/zustand/blob/main/docs/guides/typescript.md#basic-usage)

**Cons** 
- Manually defined type signature `set: SetState<typeof PostStore>` - still better than having to update the store type on each change

If in principle you agree it would be a good addition, I can open a PR. Thank you for your attention!

**Comment by dai-shi** [comment]
I think the major pain is slice pattern with middleware. It would be nice if there's a way to mitigate it.
Exposing `set` type only seems suboptimal, because we have `get` and `store` too.
Other than that, I personally like it more than `combine`.
As it turns out, `combine` wouldn't be used very much and I learned many people like typing store explicitly.
So, I think the official solution would still be `StateCreator`.

btw, what is this technique called in TypeScript?

**Comment by shfrmn** [comment]
Thank you for your response, @dai-shi!

Regarding the technique, I didn't even think this would typecheck, as circular references are generally disallowed. I suspect the reason it works is because conditional types are too confusing for the compiler to be properly inspected for circular references, but that's just speculation on my side. For now this works even under TypeScript 5 in strict mode.  

Regarding `get` and `store` it would be easy to write types for them using the same pattern. I know it would potentially mean providing three type signatures manually, but I still believe this is an order of magnitude better than the current situation. My reasoning:

1. 99% of store definitions (not middleware) only need `set`, since the state can also be accessed within `set(state => ...)`.
2. Your average store in some random react app would have around a dozen fields, all of which have to be explicitly mentioned in an interface now, and it's a lot more than three (for `set`, `get` and `store`).
3. Moreover, these three types are a kind of boilerplate which is easy to copy from file to file, whereas defining store type manually is a big pain point, especially for action function types.
4. There's no damage to other usage patterns if you agree to expose these types from `zustand`, people will just have a bit more freedom with regards to style.

There are two things in your response that I didn't really understand as an outsider, so could you please provide some examples?
1. Slice pattern with middleware
2. How does `StateCreator` type address the issue of automatic inference? I think it should totally continue to be used internally and for other needs, I just can't see how it can be adapted for the simplest and most widespread use case of implicitly typing a simple store.

Please let me know if you would consider a PR where I'd introduce `set`, `get` and `store` conditional types.

**Reply by dai-shi** [reply]
> as circular references are generally disallowed.

Is this a hack then?

> Slice pattern with middleware

https://github.com/pmndrs/zustand/blob/e489a63513509b62d33e5580d5541680dbcbffb9/docs/guides/slices-pattern.md
and https://github.com/pmndrs/zustand/blob/e489a63513509b62d33e5580d5541680dbcbffb9/docs/guides/typescript.md#slices-pattern

Even without slice pattern, this technique doesn't work with middleware, right?

> How does StateCreator type address the issue of automatic inference?

It doesn't, but at the same time, many people seem to prefer explicit typing.

> if you would consider a PR

I wouldn't at the moment, unless maybe @devanshj strongly supports it. (I would drop `combine` in such a case. That's my motivation for this.)

**Reply by shfrmn** [reply]
Hey @dai-shi


> Is this a hack then?

This would be a matter of opinion. It _can_ legally work because the parameter type only references the return type of the function and not itself, so in that sense it's not a circular dependency. I'm not familiar enough with the inner workings of TypeScript compiler to know what logic inference actually follows.

Since it's a [chicken and egg problem](https://github.com/pmndrs/zustand/blob/e489a63513509b62d33e5580d5541680dbcbffb9/docs/guides/typescript.md?plain=1#L47), as you mentioned in the docs, the cycle can only be broken with this type of technique that would allow to infer the relevant part of the function type signature.

> It doesn't, but at the same time, many people seem to prefer explicit typing.

They can continue casting types explicitly, the change I'm suggesting wouldn't prevent them from doing so.

> Even without slice pattern, this technique doesn't work with middleware, right?

I'm not sure what you mean by "doesn't work with middleware", why you think it wouldn't work and why it's relevant for the simplest use case. As for slice pattern, it works fine, just need to pass parameters separately (or continue to use `StateCreator` explicitly).

```typescript
function Store1(set: SetState<typeof Store1>) {
  // ...
}

function Store2(set: SetState<typeof Store2>) {
  // ...
}

function CombinedStore(set: SetState<typeof CombinedStore>) {
  const state1 = Store1(set)
  const state2 = Store2(set)
  return {...state1, ...state2}
}

export const useCombinedStore = create(CombinedStore)
```

My suggestion was to simply introduce helper type signatures (`SetState`, `GetState` and `Store`) to **only** serve the simple use case (as I demonstrated), without making any changes to how things are built within `zustand` or the middleware. Those would continue to use `StateCreator` type explicitly. This way it would be a non-breaking minor version change that would immediately benefit a lot of people and wouldn't require any additional work from you or the community.

All I'm trying to do here is to suggest an obvious improvement for DX, which many TypeScript users find suboptimal. I thought you'd be interested in this, especially since it's a non-breaking additive change. If neither you or @devanshj are interested, feel free to close this discussion and I'll just publish a package for anyone who has a stake in this.

**Reply by dai-shi** [reply]
> I'm not sure what you mean by "doesn't work with middleware"

because middleware can change the `setState` type.

> I thought you'd be interested in this, especially since it's a non-breaking additive change.

I am  interested, but the decision to include things isn't that simple.

> I'll just publish a package for anyone who has a stake in this.

I second the idea. Because if we know many people use it, then that's a motivation for the decision.

**Comment by shfrmn** [comment]
Posting at the top level for better discoverability. I have published these types as a separate package for anyone who needs it. You can find detailed usage instructions in the [README](https://github.com/shfrmn/zustand-types#readme).

Short demo:


```typescript
import {create} from "zustand"
import {SetState} from "zustand-types"

function MyStore(set: SetState<typeof MyStore>) {
  return {
    pets: [] as string[],
    //                            state: type correctly inferred
    addPet: (pet: string) => set((state) => [...state.pets, pet]),
  }
}

//           useStore: type correctly inferred
export const useStore = create(MyStore)
```

Install it as a dev dependency (`zustand` must be installed separately)
```bash
npm install -D zustand-types
# or
yarn add -D zustand-types
```
- source: https://www.github.com/shfrmn/zustand-types
- npm package: https://www.npmjs.com/package/zustand-types
